General Report Interpretation Guide
===================================

This is meant to be a description of all the output statistics that are produced by the EAGER ReportTable module and will therefore be updated once new statistics are added in the upcoming future.


Sample Number
-------------

The number of the sample in this project run.

Sample name
-----------

The name or ID of the sample.

# of raw reads after C&M prior mapping
--------------------------------------

This is the number of raw reads after clipping (and merging if this has been applied). This included both unmerged and merged reads, if you selected to only use merged reads for downstream analysis, look in the column
'merged reads' for your number of reads that went into the downstream analysis.


# of merged reads
-----------------

The number of merged reads, merged by Clip & Merge.


% merged reads
--------------

The percentage of merged reads, generated by applying

.. math::

   # of merged reads / # of raw reads in total

# mapped reads prior RMDup
--------------------------

The number of reads that were mapping to the provided reference genome prior to applying duplication removal either using DeDup or MarkDuplicates.


# mapped reads prior RMDup QF
-----------------------------

The number of reads that were mapping to the provided reference genome prior to applying duplication removal, also including a potentially applied quality filtering on BAM conversion.


# of Duplicates removed
-----------------------

The number of duplicates removed by the duplication removal tool.


Mapped Reads after RMDup
------------------------

The number of mapped reads after duplicate removal was applied.


Endogenous DNA (%)
------------------

The % endogenous DNA content in the selected sample.


Cluster Factor
--------------

An informative measure to determine the complexity of the underlying sample. A good cluster factor of approximately 1.0 is a sign of a high number of unique reads in the respective sample, whereas a high cluster factor
could be seen as a measure to not invest more in sequencing.

Mean Coverage
-------------

The average coverage on the provided reference genome.

std. dev. Coverage
------------------

The standard deviation of the average coverage on the provided reference genome.

Coverage >= 1X
--------------

The % of the genome covered with greater or equal than 1x coverage.

Coverage >= 2X
--------------

The % of the genome covered with greater or equal than 2x coverage.


Coverage >= 3X
--------------

The % of the genome covered with greater or equal than 3x coverage.


Coverage >= 4X
--------------

The % of the genome covered with greater or equal than 4x coverage.


Coverage >= 5X
--------------

The % of the genome covered with greater or equal than 5x coverage.


# SNPs
------

The number of SNPs found in the finally generated consensus sequence using VCF2Genome, applying the filter criteria defined in this tools as well.

AVG Coverage on mitochondrium
-----------------------------

The average coverage on the specified mitochondrial genome.

Initial cont est
----------------

The initial average contamination estimate provided by schmutzi's contDeam method.

Initial cont est low
--------------------

The initial lower 95% CI estimate on schmutzis contdeam method.

Initial cont est high
---------------------

The initial higher 95% CI estimate on schmutzis contdeam method.


Final cont est
--------------

The average contamination estimate based on the mtCont method in schmutzi.

Final cont est low
------------------

The lower 95% CI estimate on schmutzis mtCont method.


Final cont est high
-------------------

The higher 95% CI estimate on schmutzis mtCont method.


GC content
----------

The GC content of the respective sample.

# of reads on mitochondrium
---------------------------

The number of reads on the mitochondrial genome.

MT/NUC Ratio
------------

The ratio between mt and autosomal reads, calculated as follows:

.. math::

   # of mt reads / # of nuclear reads


DMG 1st Base 3'
---------------

The damage on the 1st base on the 3' end of the merged reads.


DMG 2nd Base 3'
---------------

The damage on the 2nd base on the 3' end of the merged reads.


DMG 1st Base 5'
---------------

The damage on the 1st base on the 5' end of the merged reads.


DMG 2nd Base 5'
---------------

The damage on the 2nd base on the 3' end of the merged reads.


average fragment length
-----------------------

The average fragment length of your samples reads.